catalog/catalog.go

```diff
@@ -0,0 +1,1099 @@
+package catalog
+
+import (
+	"context"
+	"crypto"
+	_ "crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/treeverse/lakefs/config"
+	"github.com/treeverse/lakefs/db"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/treeverse/lakefs/graveler/committed"
+	"github.com/treeverse/lakefs/graveler/ref"
+	"github.com/treeverse/lakefs/graveler/sstable"
+	"github.com/treeverse/lakefs/graveler/staging"
+	"github.com/treeverse/lakefs/ident"
+	"github.com/treeverse/lakefs/pyramid"
+	"github.com/treeverse/lakefs/pyramid/params"
+
+	"github.com/treeverse/lakefs/block"
+
+	"github.com/treeverse/lakefs/graveler"
+	"github.com/treeverse/lakefs/logging"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// hashAlg is the hashing algorithm to use to generate graveler identifiers.  Changing it
+// causes all old identifiers to change, so while existing installations will continue to
+// function they will be unable to re-use any existing objects.
+const hashAlg = crypto.SHA256
+
+type Path string
+
+type EntryRecord struct {
+	Path Path
+	*Entry
+}
+
+type EntryListing struct {
+	CommonPrefix bool
+	Path
+	*Entry
+}
+
+type EntryDiff struct {
+	Type  graveler.DiffType
+	Path  Path
+	Entry *Entry
+}
+
+type EntryIterator interface {
+	Next() bool
+	SeekGE(id Path)
+	Value() *EntryRecord
+	Err() error
+	Close()
+}
+
+type EntryListingIterator interface {
+	Next() bool
+	SeekGE(id Path)
+	Value() *EntryListing
+	Err() error
+	Close()
+}
+
+type EntryDiffIterator interface {
+	Next() bool
+	SeekGE(id Path)
+	Value() *EntryDiff
+	Err() error
+	Close()
+}
+
+func (id Path) String() string {
+	return string(id)
+}
+
+type Store interface {
+	graveler.KeyValueStore
+	graveler.VersionController
+	graveler.Dumper
+	graveler.Loader
+	graveler.Plumbing
+}
+
+const (
+	RangeFSName     = "range"
+	MetaRangeFSName = "meta-range"
+)
+
+type Config struct {
+	Config *config.Config
+	DB     db.Database
+	LockDB db.Database
+}
+
+type Catalog struct {
+	BlockAdapter block.Adapter
+	Store        Store
+
+	log      logging.Logger
+	managers []io.Closer
+}
+
+const (
+	ListRepositoriesLimitMax = 1000
+	ListBranchesLimitMax     = 1000
+	ListTagsLimitMax         = 1000
+	DiffLimitMax             = 1000
+	ListEntriesLimitMax      = 10000
+)
+
+var ErrUnknownDiffType = errors.New("unknown graveler difference type")
+
+func New(ctx context.Context, cfg Config) (*Catalog, error) {
+	if cfg.LockDB == nil {
+		cfg.LockDB = cfg.DB
+	}
+
+	tierFSParams, err := cfg.Config.GetCommittedTierFSParams(ctx)
```

Comment:
not related to this pr - think we need to extract this one out.  The Config should hold everything needed in order to instantiate a new catalog.